widgets/saved_workflow.py

```diff
@@ -124,96 +118,110 @@ def render_saved_workflow_preview(
             )
-        else:
-            render_author_from_user(
-                published_run.last_edited_by,
-                image_size="24px",
-                text_size="0.9rem",
-                responsive=False,
+        )
+    elif published_run.last_edited_by:
+        items.append(
+            lambda: render_author_from_user(
+                published_run.last_edited_by, image_size="24px", responsive=False
             )
+        )
 
-        if published_run.run_count > 1:
-            gui.write(" • ")
-            run_count = format_number_with_suffix(published_run.run_count)
-            gui.caption(
+    if published_run.run_count > 1:
+        run_count = format_number_with_suffix(published_run.run_count)
+        items.append(
+            lambda: gui.write(
                 f"{icons.run} {run_count} runs",
                 unsafe_allow_html=True,
-                style={"fontSize": "0.9rem"},
                 className="text-dark",
             )
+        )
+
+    updated_at = published_run.saved_run.updated_at or ""
+    if updated_at and isinstance(updated_at, datetime.datetime):
+        items.append(lambda: gui.write(f"{get_relative_time(updated_at)}"))
+
+    render_items_with_separator(
+        items,
+        className="render_example_author_meta",
+        style={"fontSize": "0.9rem"},
+    )
 
-        if published_run.last_edited_by:
-            gui.write(" • ")
 
-        if updated_at and isinstance(updated_at, datetime.datetime):
-            gui.caption(
-                f"{get_relative_time(updated_at)}",
-                style={"fontSize": "0.9rem"},
-                className="container-margin-reset",
+def render_change_notes_and_share_badge(
+    published_run: PublishedRun,
+    hide_version_notes: bool = False,
+    hide_visibility_pill: bool = False,
+    **props,
+):
+    version = published_run.versions.latest()
+    items = []
+    if not hide_version_notes and version.change_notes:
+        items.append(
+            lambda: gui.caption(
+                f"{icons.notes} {html.escape(version.change_notes)}",
+                unsafe_allow_html=True,
+                line_clamp=1,
+                lineClampExpand=False,
             )
+        )
+    if not hide_visibility_pill:
+        items.append(
+            lambda: gui.caption(
+                published_run.get_share_badge_html(), unsafe_allow_html=True
+            )
+        )
+    render_items_with_separator(items, **props, style={"fontSize": "0.9rem"})
+
+
+def render_items_with_separator(
+    items: typing.List[typing.Callable | None], *, separator: str = " • ", **props
+):
+    className = "d-flex align-items-center container-margin-reset gap-2 " + (
+        props.pop("className", "")
+    )
+    with gui.div(className=className, **props):
+        items = filter(None, items)
+        for i, item in enumerate(items):
+            if i != 0:
+                gui.write(separator)
+            item()
```

Comment:
@devxpy thoughts on this abstraction?

it was getting tedious to put separators. it was also muddying the core logic in rendering functions. i'm not very happy about the `lambda`s though.

Reply:
just use css